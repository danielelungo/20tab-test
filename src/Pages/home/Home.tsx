import { FC, useEffect, useState } from "react"
import { useNavigate } from "react-router-dom"
import { useFetch } from "../../Hooks/useFetch/useFetch"
import { Todo } from "../../Types/Todo"

const PLACEHOLDER_URL = "https://jsonplaceholder.typicode.com/todos"

const Home: FC = () => {
  const navigate = useNavigate()
  const [todos, setTodos] = useState<Todo[]>([])
  const [search, setSearch] = useState("")

  const { results, loading, error } = useFetch(PLACEHOLDER_URL)

  useEffect(() => {
    if (results.length) {
      setTodos(results.slice(0, 10))
      //.slice(0, 10)
    }
  }, [results])

  const handleOnChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value)
  }

  const filterOnINput = (todo: Todo) => todo.title.toLowerCase().includes(search.toLowerCase())

  const handleOnClickTodo = (id: number) => {
    navigate(`/todo/${id}`)
    //router.push("/todo/" + id)
  }

  const handleOnClickDelete = (id: number) => {
    setTodos(todos.filter((item) => item.id !== id))
  }

  return (
    <div>
      <main>
        <input
          value={search}
          onChange={handleOnChangeInput}
          placeholder="Search todo..."
          aria-label="todo-input"
        />
        {loading ? (
          <div>loading..</div>
        ) : error ? (
          <div>{error}</div>
        ) : (
          todos.filter(filterOnINput).map(({ id, title, completed }) => (
            <div key={id}>
              <div onClick={() => handleOnClickTodo(id)}>
                <p>{title}</p>
                {completed ? (
                  <span role="img" aria-label="emote">
                    âœ…
                  </span>
                ) : null}
              </div>
              <button data-testid={`close-btn-${id}`} onClick={() => handleOnClickDelete(id)}>
                Delete
              </button>
            </div>
          ))
        )}
      </main>
    </div>
  )
}

export default Home
