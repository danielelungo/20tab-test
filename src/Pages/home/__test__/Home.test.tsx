/* eslint-disable testing-library/prefer-query-by-disappearance */
import { render, screen, fireEvent, waitForElementToBeRemoved } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"
import renderer from "react-test-renderer"
import "@testing-library/jest-dom"
import Home from "../Home"
import { mockData } from "../../../__mock__/mockData"

const MockHome = () => {
  return (
    <BrowserRouter>
      <Home />
    </BrowserRouter>
  )
}

describe("Home Componennt", () => {
  test("match Home Page SnapShot", () => {
    const tree = renderer.create(<MockHome />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  describe("input", () => {
    test("input is in the page", () => {
      render(<MockHome />)
      const inputElement = screen.getByPlaceholderText(/search todo.../i)
      expect(inputElement).toBeInTheDocument()
    })
    test("input works", () => {
      render(<MockHome />)
      const inputElement = screen.getByLabelText("todo-input") as HTMLInputElement
      fireEvent.change(inputElement, { target: { value: "soap" } })
      expect(inputElement.value).toBe("soap")
    })
  })

  describe("todo list test", () => {
    test("should show title of todos", async () => {
      render(<MockHome />)
      await waitForElementToBeRemoved(() => screen.getByText(/loading/i))
      mockData.forEach((d) => expect(screen.getByText(d.title)).toBeInTheDocument())
    })
    test("remove todo from list", async () => {
      render(<MockHome />)
      await waitForElementToBeRemoved(() => screen.getByText(/loading/i))
      fireEvent.click(screen.getByTestId("close-btn-2"))
      expect(screen.queryByText(/quis ut nam facilis et officia qui/i)).not.toBeInTheDocument()
    })
  })
})
