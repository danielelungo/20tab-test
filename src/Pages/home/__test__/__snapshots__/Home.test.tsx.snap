// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Componennt match Home Page SnapShot 1`] = `
<div>
  <main>
    <input
      aria-label="todo-input"
      onChange={[Function]}
      placeholder="Search todo..."
      value=""
    />
    <div>
      loading..
    </div>
  </main>
</div>
`;
